cmd/snap/cmd_download.go

```diff
@@ -170,3 +159,142 @@ func (x *cmdDownload) Execute(args []string) error {
 
 	return nil
 }
+
+func (x *cmdDownload) downloadViaSnapd(snapName string, rev snap.Revision) error {
+	opts := &client.SnapOptions{
+		Channel:   x.Channel,
+		CohortKey: x.CohortKey,
+		Revision:  rev.String(),
+	}
+	fname, stream, err := x.client.Download(snapName, opts)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(Stdout, i18n.G("Fetching snap %q\n"), snapName)
+
+	if x.Basename != "" {
+		fname = x.Basename + ".snap"
+	}
+	downloadPath := filepath.Join(x.TargetDir, fname)
+	if err := os.MkdirAll(filepath.Dir(downloadPath), 0755); err != nil {
+		return err
+	}
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if _, err := io.Copy(f, stream); err != nil {
```

Comment:
this gives you no progress bar (you'll have to tweak `client.Download` to get the download size here! (it's `rsp.ContentLength`, right next to `rsp.Body` one layer up; that omission and the loss of 'resume' I should have spotted earlier, sorry), and should also be doing the sha3 checksumming.
Look at store's Download to see how to set up a multiwriter for this
Look at wait's signal handling to see how to set things up for the progress bar to dtrt in the face of signals

Reply:
Also, the store told you what the sha3 *should* be, and we discarded that information. Maybe it should be set in a header in the response, in `store.DownloadStream`, so that the client can pick it up and check it before hitting the assertions db.

Reply:
We *do* have the size! `response.fileStream` seems to DTRT with it.

Reply:
@chipaca @mvo idea, maybe expose the store.go downloadImpl helper is some form ?

Reply:
possible there's a way to split out the bits we want to reuse from `downloadImpl` into their own helper that can then be exposed, but it's not immediately obvious to me today